tests/plugins/clipboard/dropImage.js

```diff
@@ -0,0 +1,169 @@
+
+		'test drop .jpeg image': function() {
+			var dropEvt = bender.tools.mockDropEvent(),
+				imageType = 'image/jpeg',
+				expected = '<p class="p">Paste image here:<img src="data:image/jpeg;base64,fileMockBase64=" /></p>';
+
+			FileReader.setFileMockType( imageType );
+			FileReader.setReadResult( 'load' );
+
+			setHtmlWithSelection( this.editor, '<p class="p">Paste image here:^</p>' );
+			this.assertDropImage( dropEvt, imageType, expected, {
+				dropContainer: this.editor.editable().findOne( '.p' ).getChild( 0 ),
+				dropOffset: 17
+			} );
+		},
+
+		'test drop .gif image': function() {
+			var dropEvt = bender.tools.mockDropEvent(),
+				imageType = 'image/gif',
+				expected = '<p class="p">Paste image here:<img src="data:image/gif;base64,fileMockBase64=" /></p>';
+
+			FileReader.setFileMockType( imageType );
+			FileReader.setReadResult( 'load' );
+
+			setHtmlWithSelection( this.editor, '<p class="p">Paste image here:^</p>' );
+			this.assertDropImage( dropEvt, imageType, expected, {
+				dropContainer: this.editor.editable().findOne( '.p' ).getChild( 0 ),
+				dropOffset: 17
+			} );
+		},
+
+		'test drop unsupported image type': function() {
+			var dropEvt = bender.tools.mockDropEvent(),
+				imageType = 'application/pdf',
+				expected = '<p class="p">Paste image here:</p>';
+
+			FileReader.setFileMockType( imageType );
+			FileReader.setReadResult( 'load' );
+
+			setHtmlWithSelection( this.editor, '<p class="p">Paste image here:^</p>' );
+			this.assertDropImage( dropEvt, imageType, expected, {
+				dropContainer: this.editor.editable().findOne( '.p' ).getChild( 0 ),
+				dropOffset: 17
+			} );
+		},
+
+		'test abort drop': function() {
+			var dropEvt = bender.tools.mockDropEvent(),
+				imageType = 'image/png',
+				expected = '<p class="p">Paste image here:</p>';
+
+			FileReader.setFileMockType( imageType );
+			FileReader.setReadResult( 'abort' );
+
+			setHtmlWithSelection( this.editor, '<p class="p">Paste image here:^</p>' );
+			this.assertDropImage( dropEvt, imageType, expected, {
+				dropContainer: this.editor.editable().findOne( '.p' ).getChild( 0 ),
+				dropOffset: 17
+			} );
+		},
+
+		'test failed drop': function() {
+			var dropEvt = bender.tools.mockDropEvent(),
+				imageType = 'image/gif',
+				expected = '<p class="p">Paste image here:</p>';
+
+			FileReader.setFileMockType( imageType );
+			FileReader.setReadResult( 'error' );
+
+			setHtmlWithSelection( this.editor, '<p class="p">Paste image here:^</p>' );
+			this.assertDropImage( dropEvt, imageType, expected, {
+				dropContainer: this.editor.editable().findOne( '.p' ).getChild( 0 ),
+				dropOffset: 17
+			} );
+		},
+
+		assertDropImage: function( evt, type, expected, config ) {
+			var dropTarget = CKEDITOR.plugins.clipboard.getDropTarget( this.editor ),
+				range = new CKEDITOR.dom.range( this.editor.document );
+
+			if ( CKEDITOR.env.gecko ) {
+				range.setStart( config.dropContainer, config.dropOffset );
+				evt.testRange = range;
+			}
```

Comment:
Why is it needed for Gecko?

Reply:
Because there was a problem with getting a range on FF .